addons/mail/models/discuss/discuss_channel.py

```diff
@@ -1087,33 +1088,66 @@ def _types_allowing_seen_infos(self):
     def channel_fetched(self):
         """ Broadcast the channel_fetched notification to channel members
         """
-        for channel in self:
-            if not channel.message_ids.ids:
-                return
-            # a bit not-modular but helps understanding code
-            if channel.channel_type not in {'chat', 'whatsapp'}:
-                return
-            last_message_id = channel.message_ids.ids[0] # zero is the index of the last message
-            member = self.env['discuss.channel.member'].search([('channel_id', '=', channel.id), ('partner_id', '=', self.env.user.partner_id.id)], limit=1)
-            if member.fetched_message_id.id == last_message_id:
-                # last message fetched by user is already up-to-date
-                return
-            # Avoid serialization error when multiple tabs are opened.
+        to_update = []  # (member_id, last_message_id)
+        notifications = []  # channel, notification_type, message
+        partner_id = self.env.user.partner_id.id
+
+        # Avoid serialization error when multiple tabs are opened.
+        # In order to completely avoid any issues with concurrency, the very first thing we do on
+        # the cursor needs to be a SELECT FOR UPDATE
+        with self.env.registry.cursor() as cr:
+            self = self.with_env(self.env(cr=cr))
+            member_query = self.env['discuss.channel.member']._search([
+                ('channel_id', 'in', self.ids),
+                ('partner_id', '=', partner_id),
+            ])
+            all_member_result = self.env.execute_query(SQL("""
+                SELECT id, fetched_message_id, channel_id
+                  FROM discuss_channel_member
+                 WHERE id IN (%s)
+                   FOR NO KEY UPDATE SKIP LOCKED
+            """, member_query.subselect()))
+            current_member_data = {
+                channel_id: (member_id, message_id)
+                for member_id, message_id, channel_id in all_member_result
+            }
```

Comment:
I wish there was a more simply syntax with the ORM for this use case.

Reply:
Me too :laughing: 

Maybe a parameter in `search`, but it would still require to be used correctly and to lock at the right moment, which is not an easy task and needs good testing.

Reply:
The best I thought of was this, but I don't really like it and it has a lot of restrictions.

```diff
diff --git a/addons/mail/models/discuss/discuss_channel.py b/addons/mail/models/discuss/discuss_channel.py
index d75e452deee9..4d57d2f79d89 100644
--- a/addons/mail/models/discuss/discuss_channel.py
+++ b/addons/mail/models/discuss/discuss_channel.py
@@ -1100,16 +1100,11 @@ class Channel(models.Model):
         # In order to completely avoid any issues with concurrency, the very first thing we do on
         # the cursor needs to be a SELECT FOR UPDATE
         member_query = self.env['discuss.channel.member']._search([('channel_id', 'in', self.ids), ('is_self', '=', True)])
+        member_query._for = 'NO KEY UPDATE SKIP LOCKED'
         with self.env.registry.cursor() as cr:
             self = self.with_env(self.env(cr=cr))
-            to_update = self.env.execute_query(SQL("""
-                SELECT id
-                  FROM discuss_channel_member
-                 WHERE id IN (%s)
-                   FOR NO KEY UPDATE SKIP LOCKED
-            """, member_query.subselect()))
             notifications = []  # (channel, notification_type, message)
-            for member in self.env['discuss.channel.member'].browse([r[0] for r in to_update]):
+            for member in self.env['discuss.channel.member']._fetch_query(member_query, []):
                 if member.channel_id.channel_type not in ('chat', 'whatsapp'):
                     continue
                 last_message = member.channel_id._fetch_last_message()
diff --git a/odoo/tools/query.py b/odoo/tools/query.py
index 83ef66918dd6..938abda4cc6d 100644
--- a/odoo/tools/query.py
+++ b/odoo/tools/query.py
@@ -70,6 +70,7 @@ class Query:
         self._order = None
         self.limit = None
         self.offset = None
+        self._for = None

         # memoized result
         self._ids = None
@@ -175,15 +176,28 @@ class Query:
     def select(self, *args: str | SQL) -> SQL:
         """ Return the SELECT query as an ``SQL`` object. """
         sql_args = map(SQL, args) if args else [SQL.identifier(self.table, 'id')]
-        return SQL(
-            "%s%s%s%s%s%s",
-            SQL("SELECT %s", SQL(", ").join(sql_args)),
+        select = SQL("SELECT %s", SQL(", ").join(sql_args))
+        from_where = SQL(
+            "%s%s",
             SQL(" FROM %s", self.from_clause),
             SQL(" WHERE %s", self.where_clause) if self._where_clauses else SQL(),
+        )
+        order_limit_offset = SQL(
+            "%s%s%s",
             SQL(" ORDER BY %s", self._order) if self._order else SQL(),
             SQL(" LIMIT %s", self.limit) if self.limit else SQL(),
             SQL(" OFFSET %s", self.offset) if self.offset else SQL(),
         )
+        if self._for:
+            return SQL(
+                "%s FROM %s WHERE id IN (SELECT id %s) %s FOR %s",
+                select,
+                self.from_clause,
+                from_where,
+                order_limit_offset
+                SQL(self._for),
+            )
+        return SQL("%s%s%s", select, from_where, order_limit_offset)

     def subselect(self, *args: str | SQL) -> SQL:
         """ Similar to :meth:`.select`, but for sub-queries.
```